pulp_docker/app/models.py

```diff
@@ -1,37 +1,177 @@
+    """
+    A blob defined within a manifest.
+
+    This content has no artifacts.
+
+    Fields:
+        digest (models.CharField): The manifest digest.
+        schema_version (models.IntegerField): The docker schema version.
+        media_type (models.CharField): The manifest media type.
+
+    Relations:
+        manifests (models.ManyToManyField): Many-to-many relationship with Manifest.
+    """
+
+    TYPE = 'manifest-list'
+
+    digest = models.CharField(max_length=255)
+    schema_version = models.IntegerField()
+    media_type = models.CharField(
+        max_length=60,
+        choices=(
+            (MEDIA_TYPE.MANIFEST_LIST, MEDIA_TYPE.MANIFEST_LIST),
+        ))
+
+    manifests = models.ManyToManyField(Manifest, through='ManifestListManifest')
+
+    class Meta:
+        unique_together = ('digest',)
+
+
+class ManifestListManifest(models.Model):
+    """
+    The manifest referenced by a manifest list.
+
+    Fields:
+        architecture (models.CharField): The platform architecture.
+        variant (models.CharField): The platform variant.
+        features (models.TextField): The platform features.
+        os (models.CharField): The platform OS name.
+        os_version (models.CharField): The platform OS version.
+        os_features (models.TextField): The platform OS features.
+
+    Relations:
+        manifest (models.ForeignKey): Many-to-one relationship with Manifest.
+        manifest_list (models.ForeignKey): Many-to-one relationship with ManifestList.
+    """
+
+    architecture = models.CharField(max_length=255)
+    os = models.CharField(max_length=255)
+    os_version = models.CharField(max_length=255)
+    os_features = models.TextField(default='', blank=True)
+    features = models.TextField(default='', blank=True)
+    variant = models.CharField(max_length=255)
+
+    manifest = models.ForeignKey(
+        Manifest, related_name='manifests', on_delete=models.CASCADE)
+    manifest_list = models.ForeignKey(
+        ManifestList, related_name='manifest_lists', on_delete=models.CASCADE)
+
+
+class Tag(Content):
+    """
+    A docker tag.
+
+    Each tag will reference either a Manifest or a ManifestList.
+    A repository may contain tags with duplicate names provided each tag references
+    a different type of object (Manifest|ManifestList).  This uniqueness is enforced
+    programmatically.
+
+    This content has no artifacts.
+
+    Fields:
+        name (models.CharField): The tag name.
+
+    Relations:
+        manifest (models.ForeignKey): A referenced Manifest.
+        manifest_list (models.ForeignKey): A referenced ManifestList.
+
+    """
+
+    TYPE = 'tag'
+
+    name = models.CharField(max_length=255, db_index=True)
+
+    manifest = models.ForeignKey(
+        Manifest, null=True, related_name='tags', on_delete=models.CASCADE)
+    manifest_list = models.ForeignKey(
+        ManifestList, null=True, related_name='tags', on_delete=models.CASCADE)
+
+    class Meta:
```

Comment:
Nonblocking question: Is there a way to enforce that a specific tag cannot reference both a manifest and a manifest list?

Reply:
@dkliban, do you know of a way that is compatible with bulk_create()?